sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -359,7 +359,7 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
   // TimestampConverter
   private[this] def castToTimestamp(from: DataType): Any => Any = from match {
     case StringType =>
-      buildCast[UTF8String](_, utfs => DateTimeUtils.stringToTimestamp(utfs, timeZone).orNull)
+      buildCast[UTF8String](_, s => DateTimeUtils.stringToTimestamp(s.trim(), timeZone).orNull)
```

Comment:
What about changing `stringToDate` and `stringToTimestamp` instead?
Those functions are used only in `Cast` and they already handle `null`cases, too.
I didn't look at the detail of this PR. The change looks a little less robust when `s` is `null`.

Reply:
cc @gatorsmile

Reply:
How about change `stringToDate` to `trimStringToDate` and update `trimStringToDate` to:
![image](https://user-images.githubusercontent.com/5399861/48036353-ec49a100-e1a2-11e8-80e6-b52b4a007493.png)

Reply:
Ur, I'd like not to rename it. One line function document will suffice.